man/systemd.network.xml

```diff
@@ -859,6 +859,14 @@ Table=1234</programlisting></para>
           <citerefentry><refentrytitle>networkd.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>.
           </para>
 
+          <para>Note, unlike <varname>IPv4Forwarding=</varname>, enabling per-interface
+          <varname>IPv6Forwarding=</varname> on two or more interfaces <emphasis>DOES NOT</emphasis> make IPv6
+          packets forwarded within the interfaces. This setting just controls the per-interface sysctl value,
+          and the sysctl is not directly connected to if packets are forwarded. To make IPv6 packets
+          forwarded, the global setting in
+          <citerefentry><refentrytitle>networkd.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>
+          needs to be enabled.</para>
```

Comment:
this puzzles me. first of all, i have no idea *what* this does. It tells me what it doesn't do, and that's certainly useful info, but options really should give me a hint of what they actually do in the first place.

i don't grok the kernel's own docs on this, this just leaves me head scratching...

if this doesn't do anything useful we probbaly shouldn't have this option at all.

does anyone grok what it actually does?

Reply:
If the per-interface option is set, the interface behaves like an interface which is part of a router, which means (from the sysctl docs):
```
        1. IsRouter flag is set in Neighbour Advertisements.
	2. Router Solicitations are not sent unless accept_ra is 2.
	3. Router Advertisements are ignored unless accept_ra is 2.
	4. Redirects are ignored.
```

To me this is pretty clear, but please ping me if you need further explanation.

Reply:
so, of those 4 options, how many are actually relevant in networkd context, given we to RA on our own?

Reply:
Unfortunately, iiuc, the kernel uses the flag at several more places, and that's not documented anywhere. So, I intentionally do not list what the kernel does with the flag. If my understanding is correct, the kernel doc needs to be updated at first. The first sentence suggested by @arvidjaar is slightly ambiguous but should be correct and the best one I can add there at least now.

Reply:
Just for posterity, this is from the `systemd.network` man:
```
Note that kernel's implementation of the IPv6 RA protocol is always disabled, regardless of this setting. If this option is enabled, a userspace implementation of the
IPv6 RA protocol is used, and the kernel's own implementation remains disabled, since systemd-networkd needs to know all details supplied in the advertisements, and these
are not available from the kernel if the kernel's own implementation is used.
```

Given that the kernel's RA protocol stack is disabled anyways, we can just omit this option per-interface.

Reply:
@freswa IsRouter flag in NA is important. networkd does not handle NS, and does not send NA.

Reply:
I ran quickly through RFC4861 but I fail to find anything disallowing routers to to send RS themselves. Though doing so doesn't make any sense to me either.

I think `1` and `2` is kinda important for us to set while `3` and `4` are handled by networkd.

Reply:
> Unfortunately, iiuc, the kernel uses the flag at several more places, and that's not documented anywhere

Kernel is using the value of `ipv6.conf.`**`ALL`**`.forwarding` to decide whether packages should be **forwarded** between interfaces (technically setting this value to `0` restricts selection of routing entries to the directly accessible ones). Setting the `ipv6.conf.all.forwarding` also resets the values of `ipv6.conf.`*`interface`*`.forwarding` that define address configuration behavior listed here. Per-interface flag does **not** affect **forwarding**. If I am wrong - demonstrate the configuration where setting per-interface `forwarding` to `0` actually disables forwarding to/from this interface.

Yes, as opposed to `ipv4` case the `ipv6.conf.all.forwarding` is not just shortcut for setting the value for all interfaces. It it is checked explicitly, not only when setting the value.

Reply:
> Per-interface flag does **not** affect **forwarding**.

To demonstrate (there are two namespaces `n1` and `n2`).

n1:
```
n1:~ # ip a show dev veth-host
2: veth-host@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether ca:ad:80:ee:7e:37 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fd21::2/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::c8ad:80ff:feee:7e37/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
n1:~ # ip -6 r
fd21::/64 dev veth-host proto kernel metric 256 pref medium
fe80::/64 dev veth-host proto kernel metric 256 pref medium
default via fd21::1 dev veth-host metric 1024 pref medium
n1:~ # 
```
n2:
```
n2:~ # ip a show dev veth-host
2: veth-host@if5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 1e:18:90:8f:89:60 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fd22::2/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::1c18:90ff:fe8f:8960/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
n2:~ # ip -6 r
fd22::/64 dev veth-host proto kernel metric 256 pref medium
fe80::/64 dev veth-host proto kernel metric 256 pref medium
default via fd22::1 dev veth-host metric 1024 pref medium
n2:~ # 
```

host (gateway)
```
tw:~ # ip a show dev veth-n1
4: veth-n1@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether b2:00:77:7e:b5:48 brd ff:ff:ff:ff:ff:ff link-netns n1
    inet6 fd21::1/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::b000:77ff:fe7e:b548/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
tw:~ # ip a show dev veth-n2
5: veth-n2@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether e6:8b:3b:ff:76:ee brd ff:ff:ff:ff:ff:ff link-netns n2
    inet6 fd22::1/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::e48b:3bff:feff:76ee/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
tw:~ # grep . /proc/sys/net/ipv6/conf/{all,veth*}/forwarding
/proc/sys/net/ipv6/conf/all/forwarding:1
/proc/sys/net/ipv6/conf/veth-n1/forwarding:0
/proc/sys/net/ipv6/conf/veth-n2/forwarding:0
tw:~ # 
```
Although `forwarding` is set to `0` on both interfaces, it perfectly forwards packets between them.
```
n1:~ # ping -6 -c 2 fd22::2
PING fd22::2 (fd22::2) 56 data bytes
64 bytes from fd22::2: icmp_seq=1 ttl=63 time=0.121 ms
64 bytes from fd22::2: icmp_seq=2 ttl=63 time=0.049 ms

--- fd22::2 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1013ms
rtt min/avg/max/mdev = 0.049/0.085/0.121/0.036 ms
n1:~ # 
```
```
n2:~ # ping -6 -c 2 fd21::2
PING fd21::2 (fd21::2) 56 data bytes
64 bytes from fd21::2: icmp_seq=1 ttl=63 time=0.333 ms
64 bytes from fd21::2: icmp_seq=2 ttl=63 time=0.050 ms

--- fd21::2 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1006ms
rtt min/avg/max/mdev = 0.050/0.191/0.333/0.141 ms
n2:~ # 
```

To disable forwarding one needs to set `all/forwarding` to `0`.
```
tw:~ # echo 0 > /proc/sys/net/ipv6/conf/all/forwarding
tw:~ # grep . /proc/sys/net/ipv6/conf/{all,veth*}/forwarding
/proc/sys/net/ipv6/conf/all/forwarding:0
/proc/sys/net/ipv6/conf/veth-n1/forwarding:0
/proc/sys/net/ipv6/conf/veth-n2/forwarding:0
tw:~ # 
n1:~ # ping -6 -c 2 fd22::2
PING fd22::2 (fd22::2) 56 data bytes

--- fd22::2 ping statistics ---
2 packets transmitted, 0 received, 100% packet loss, time 1007ms

n1:~ # 
n2:~ # ping -6 -c 2 fd21::2
PING fd21::2 (fd21::2) 56 data bytes

--- fd21::2 ping statistics ---
2 packets transmitted, 0 received, 100% packet loss, time 1013ms

n2:~ # 
```
Setting per-interface `forwarding` to `1` does not change anything.
```
tw:~ # echo 1 > /proc/sys/net/ipv6/conf/veth-n1/forwarding
tw:~ # echo 1 > /proc/sys/net/ipv6/conf/veth-n2/forwarding
tw:~ # grep . /proc/sys/net/ipv6/conf/{all,veth*}/forwarding
/proc/sys/net/ipv6/conf/all/forwarding:0
/proc/sys/net/ipv6/conf/veth-n1/forwarding:1
/proc/sys/net/ipv6/conf/veth-n2/forwarding:1
tw:~ # 
n1:~ # ping -6 -c 2 fd22::2
PING fd22::2 (fd22::2) 56 data bytes

--- fd22::2 ping statistics ---
2 packets transmitted, 0 received, 100% packet loss, time 1024ms

n1:~ # 
n2:~ # ping -6 -c 2 fd21::2
PING fd21::2 (fd21::2) 56 data bytes

--- fd21::2 ping statistics ---
2 packets transmitted, 0 received, 100% packet loss, time 1016ms

n2:~ # 
```




> The first sentence suggested by @arvidjaar is slightly ambiguous but should be correct

As demonstrated, it is absolutely incorrect. This setting has nothing to do with forwarding at all.

Reply:
> > The first sentence suggested by @arvidjaar is slightly ambiguous but should be correct
> 
> As demonstrated, it is absolutely incorrect. This setting has nothing to do with forwarding at all.

Now the doc does not say it controls forwarding. It may still misleading or hard to understand, but should be correct, at least, not incorrect.